### PR TITLE
REST to GraphQL.

### PR DESCRIPTION
**Pull request by AndjelaBoltaKovacic**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "cSpell.words": ["asynciterable", "Imgs"]
```
Comment (omega-interviews) [inline .vscode/settings.json:2]: .vscode setting should be in .gitignore

File: src/app/graphql.module.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { NgModule } from '@angular/core';
+import { APOLLO_OPTIONS } from 'apollo-angular';
+import { ApolloClientOptions, InMemoryCache } from '@apollo/client/core';
+import { HttpLink } from 'apollo-angular/http';
+
+const uri = 'https://api.spacex.land/graphql/'; // <-- add the URL of the GraphQL server here
```
Comment (omega-interviews) [inline src/app/graphql.module.ts:6]: API url should be environment variable

File: src/app/launch-details/launch-details.component.html [hunk 3]
```diff
@@ -4,24 +4,14 @@ <h3>{{ errorMessage }}</h3>
...
-          width="200px"
-        />
-        <img
-          src="/assets/images/replacement_logo.png"
-          *ngIf="launch.links.mission_patch == ''"
-          class="mission_logo"
-          width="200px"
-        />
+        <img *ngIf="launch.links.mission_patch; else backupMissionPatch" class="mission_logo"
+          src="{{ launch.links.mission_patch }}" alt="mission_patch" width="200px" />
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.html:8]: We should follow Angular's best practices and use official property binding syntax like `[src]="someImgUrl"`

File: src/app/launch-details/launch-details.component.ts [hunk 4]
```diff
@@ -1,44 +1,57 @@
...
 
   ngOnInit(): void {
-    this.route.params.subscribe((params: any) => {
-      const myParams = params[`flight_number`];
-      this.flightNumber = myParams;
-      this.getLaunch();
+    this.route.params.subscribe((parameter) => {
+      this.parameter = parameter.id;
+      this.getGraphQLLaunch();
+      console.log(this.error);
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:29]: console log

File: src/app/launch-details/launch-details.component.ts [hunk 5]
```diff
@@ -1,44 +1,57 @@
...
 })
 export class LaunchDetailsComponent implements OnInit {
-  flightNumber: number;
   launch: Launch = new Launch();
   images: string[];
   success: boolean;
   errorMessage: string;
+  noImgs: boolean;
+  name: string;
+  parameter: string;
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:20]: `parameter` is not very descriptive name for a variable. What kind of parameter? Route?

File: src/app/launch-details/launch-details.component.ts [hunk 6]
```diff
@@ -1,44 +1,57 @@
...
+  error;
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:21]: Same as previous comment :) Also type is missing.

File: src/app/launch-details/launch-details.component.ts [hunk 7]
```diff
@@ -1,44 +1,57 @@
...
     });
   }
 
-  getLaunch(): void {
-    this.service.getOne(this.flightNumber).subscribe(
-      (data) => {
-        this.flightNumber = data.flight_number;
-        this.launch = data;
+  getGraphQLLaunch(): any {
+    return this.apollo
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:34]: We could extract this as a service method.

File: src/app/launch-details/launch-details.component.ts [hunk 8]
```diff
@@ -1,44 +1,57 @@
...
+      .watchQuery<any>({
+        query: LAUNCH_QUERY,
+        errorPolicy: 'all',
+        variables: {
+          id: this.parameter,
+        },
+      })
+      .valueChanges.subscribe(({ data }) => {
+        this.launch = data.launch;
         if (this.launch.rocket.first_stage.cores[0].land_success) {
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:44]: Line 44 to 48 could be one-liner 👇 
`this.success = this.launch.rocket.first_stage.cores[0].land_success;`

File: src/app/launch-details/launch-details.component.ts [hunk 9]
```diff
@@ -1,44 +1,57 @@
...
           this.success = true;
         } else {
           this.success = false;
         }
-      },
-      (error) => (this.errorMessage = error)
-    );
+        if (this.launch.links.flickr_images.length == 0) {
```
Comment (omega-interviews) [inline src/app/launch-details/launch-details.component.ts:49]: Same as previous comment .

File: src/app/launches/launches.component.ts [hunk 10]
```diff
@@ -17,7 +18,7 @@ export class LaunchesComponent implements OnInit, AfterViewInit {
   launchList: Launch[] = [];
   filterTerm: string;
   private searchSubject = new Subject<any>();
-  noResults: boolean;
+  noResults: boolean = false;
   stillSearching: boolean;
   errorMessage: string;
   observer: IntersectionObserver;
```
Comment (omega-interviews) [inline src/app/launches/launches.component.ts:24]: Variable name. What kind of observer?

File: src/app/launches/launches.component.ts [hunk 11]
```diff
@@ -32,29 +33,41 @@ export class LaunchesComponent implements OnInit, AfterViewInit {
...
   }
 
-  getLaunches() {
-    this.service.getLaunches(this.params).subscribe(
-      (data: Launch[]) => {
-        this.launchList.push(...data);
-      },
-      (error) => (this.errorMessage = error)
-    );
+  getGrapQLLaunches(): any {
```
Comment (omega-interviews) [inline src/app/launches/launches.component.ts:52]: Extract this as a service method.

File: src/app/model/launch.model.ts [hunk 12]
```diff
@@ -2,16 +2,17 @@ import { Rocket } from './rocket.model';
...
 
 export class Launch {
-  flight_number: number;
-  mission_name: string;
-  launch_year: number;
-  launch_date_utc: string;
-  rocket: Rocket;
-  links: LaunchLinks;
-  details: string;
+  id?: any;
```
Comment (omega-interviews) [inline src/app/model/launch.model.ts:5]: We should avoid `any` types.

File: src/app/services/errorHandler.service.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { throwError } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ErrorHandlerService {
+  myQuery: any;
```
Comment (omega-interviews) [inline src/app/services/errorHandler.service.ts:9]: `any` type.

File: src/app/services/errorHandler.service.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+  constructor(public http: HttpClient) {}
+
+  errorHandler(error: any) {
```
Comment (omega-interviews) [inline src/app/services/errorHandler.service.ts:13]: `any` type

File: .vscode/settings.json (hunk 1 above)
Comment (nmitrovic-vega) [reply .vscode/settings.json:2]: .vscode setting should be in .gitignore

File: src/app/graphql.module.ts (hunk 2 above)
Comment (nmitrovic-vega) [reply src/app/graphql.module.ts:6]: API url should be environment variable

File: src/app/launch-details/launch-details.component.html (hunk 3 above)
Comment (nmitrovic-vega) [inline src/app/launch-details/launch-details.component.html:8]: We should follow Angular's best practices and use official property binding syntax like [src]="someImgUrl"

File: src/app/launch-details/launch-details.component.ts (hunk 4 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:29]: console log

File: src/app/launch-details/launch-details.component.ts (hunk 5 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:20]: parameter is not very descriptive name for a variable. What kind of parameter? Route?

File: src/app/launch-details/launch-details.component.ts (hunk 6 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:21]: Same as previous comment :) Also type is missing.

File: src/app/launch-details/launch-details.component.ts (hunk 7 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:34]: We could extract this as a service method.

File: src/app/launch-details/launch-details.component.ts (hunk 8 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:44]: Line 44 to 48 could be one-liner 👇
this.success = this.launch.rocket.first_stage.cores[0].land_success;

File: src/app/launch-details/launch-details.component.ts (hunk 9 above)
Comment (nmitrovic-vega) [reply src/app/launch-details/launch-details.component.ts:49]: Same as previous comment .

File: src/app/launches/launches.component.ts (hunk 10 above)
Comment (nmitrovic-vega) [reply src/app/launches/launches.component.ts:24]: Variable name. What kind of observer?

File: src/app/launches/launches.component.ts (hunk 11 above)
Comment (nmitrovic-vega) [reply src/app/launches/launches.component.ts:52]: Extract this as a service method.

File: src/app/model/launch.model.ts (hunk 12 above)
Comment (nmitrovic-vega) [reply src/app/model/launch.model.ts:5]: We should avoid any types.

File: src/app/services/errorHandler.service.ts (hunk 14 above)
Comment (nmitrovic-vega) [inline src/app/services/errorHandler.service.ts:13]: `any` type

File: src/app/services/errorHandler.service.ts (hunk 13 above)
Comment (nmitrovic-vega) [inline src/app/services/errorHandler.service.ts:9]: `any` type.

**Pull request merged**